contrib/babelfishpg_tsql/src/pl_comp.c

```diff
@@ -1642,6 +1667,203 @@ pltsql_param_ref(ParseState *pstate, ParamRef *pref)
+{
+
+	if (SINGLE_DB == mode)
+	{
+		if ((strlen(db_name) == 6 && (strncmp(db_name, "master", 6) == 0)) ||
+			(strlen(db_name) == 6 && (strncmp(db_name, "tempdb", 6) == 0)) ||
+			(strlen(db_name) == 4 && (strncmp(db_name, "msdb", 4) == 0)))
+		{
+			return schema_name + strlen(db_name) + 1;
+		}
+		else
+		{
+			/* all schema names are not prepended with db name on single-db */
+			return schema_name;
+		}
+	}
+	else
+	{
+		return schema_name + strlen(db_name) + 1;
+	}
+
+	return schema_name;
+}
+
+/*
+ * This function identifies and modifies geospatial function call for
+ * STX, STY, LAT, LONG functions
+ * TODO: Add handling for more such TSQL function syntax here
+ */
+static Node *
+resolve_geospatial_col_ref(ParseState *pstate, ColumnRef *cref)
+{
+	Node       *last_field;		/* represents the geospatial function name */
+	Node       *col;
+
+	/* if there is only one field then it cannot be a geospatial function call */
+	if (list_length(cref->fields) <= 1){
+		isGeospatialFunction = false;	/* resetting the global variable */
+		return NULL;
+	}
+
+	last_field = (Node *) llast(cref->fields);
+	/* 
+	 * checking if last field matches any of the geospatial functions - stx, sty, lat, long
+	 * and also checking if isGeospatialFunction variable is false to avoid entering the loop if it is a column rather than function
+	 */
+	if (IsA(last_field, String) && !isGeospatialFunction &&
+		(pg_strcasecmp(strVal(last_field), "stx") == 0 ||
+		pg_strcasecmp(strVal(last_field), "sty") == 0 ||
+		pg_strcasecmp(strVal(last_field), "lat") == 0 ||
+		pg_strcasecmp(strVal(last_field), "long") == 0))
+	{
+		FuncExpr   *funcexpr = makeNode(FuncExpr);	/* func expression to modify colref to funcref */
+		char *name = strVal(last_field);
+		Oid funcid;
+		Oid dataTypeId;
+		ColumnRef *colref = makeNode(ColumnRef);	/* new colref to avoid making changes to existing structure */
+
+		isGeospatialFunction = true;
+		colref->fields = list_copy(cref->fields);
+		/* if length is 3 then colref would have been modified by rewrite_column_refs function */
+		if(list_length(colref->fields) == 3)
+		{
+			Node	   *schema = (Node *) linitial(colref->fields);
+			char	   *cur_db = get_cur_db_name();
+			String	   *new_schema;
+
+			if (!is_shared_schema(strVal(schema)))
+			{
+				new_schema = makeString(resolve_schema_name(cur_db, strVal(schema), get_migration_mode()));
+				colref->fields = list_delete_first(colref->fields);
+				colref->fields = lcons(new_schema, colref->fields);
+			}
+		}
+		colref->fields = list_delete_last(colref->fields);	/* deletes the function name and check if the remaining colref is a valid column reference */
+		PG_TRY();
+		{
+			col = transformExpr(pstate, (Node *) colref, pstate->p_expr_kind);	/* Recursive call to check for valid column reference */
+		}
+		PG_CATCH();
+		{
+			isGeospatialFunction = false;	/* resetting the global variable */
+			return NULL;	/* if not a valid column ref */
+		}
+		PG_END_TRY();
```

Comment:
Here, we are first checking if the fields were rewritten, if Yes( i.e. list_length(colref->fields) == 3), then we are reverting that change because of our syntax using resolve_schema_name function.
Then, we are deleting the last field from the column ref and checking if the remaining colref is a valid syntax using recursive call (col = transformExpr(pstate, (Node *) colref, pstate->p_expr_kind)), if it found to be valid syntax then we will modify to funcNode else we will return NULL

Reply:
Eg: select t2.geom.STX from t2;
here since it has been modified to master_t2.geom.STX, so we will first revert it to t2.geom.STX
then, we remove last field STX, and check validity of syntax t2.geom using recursive call,
since it is a valid colRef so we modify it to FuncNode representing -> "STX(t2.geom)'